lightningd/datastore.c

```diff
@@ -298,34 +298,40 @@ static struct command_result *json_datastoreusage(struct command *cmd,
 	const char **k, **key = NULL;
 	struct db_stmt *stmt;
 	const u8 *data;
-	u64 gen;
-	u64 total = 0;
+	u64 gen, total_bytes = 0;
 
-	if (!param(cmd, buffer, params, NULL))
+	if (!param(cmd, buffer, params,
+		   p_opt("key", param_list_or_string, &key),
+		   NULL))
 		return command_param_failed();
 
+	// We ignore an empty key string or key array.
+	if (key && *key[0] == '\0')
+		key = NULL;
+
 	response = json_stream_success(cmd);
-	json_array_start(response, "usage");
-	
-	for (stmt = wallet_datastore_first(cmd, cmd->ld->wallet, key, 
+	json_object_start(response, "datastoreusage");
+	json_add_string(response, "key", datastore_key_fmt(tmpctx, key));
+
+	for (stmt = wallet_datastore_first(cmd, cmd->ld->wallet, key,
 					   &k, &data, &gen);
 	     stmt;
-	     stmt = wallet_datastore_next(cmd, cmd->ld->wallet, stmt, 
+	     stmt = wallet_datastore_next(cmd, key, stmt,
 	     				  &k, &data, &gen)) {
-		
-		json_object_start(response, NULL);
-		json_array_start(response, "key");
-		for (size_t i = 0; i < tal_count(k); i++)
-			json_add_string(response, NULL, k[i]);
-		json_array_end(response);
-		json_add_u64(response, "size", tal_bytelen(data));
-		json_object_end(response);
-		total = tal_bytelen(data) + total;
+
+		u64 self_bytes = tal_bytelen(data);
+
+		// We use the formatted datastore key to calculate the size of
+		// the key to account for the delimiter byte in the database.
+		self_bytes += strlen(datastore_key_fmt(tmpctx, k)) - 2;
```

Comment:
Doing the fmt to get the right answer seems strange here?

How about:

```
u64 self_bytes = tal_bytelen(data);
/* One byte per separator */
self_bytes += tal_count(key) - 1;
for (size_t i = 0; i < tal_count(key); i++) {
        self_bytes += strlen(key[i]);
```

Reply:
Agreed, It now is calculated form the key chars itself not the formatted string. This is straight forward and makes it even easier as it safes us a weird `-2` or `-1` to remove separators from the calculation.